components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java

```diff
@@ -580,6 +614,266 @@ protected void initiateAuthenticationRequest(HttpServletRequest request, HttpSer
         return;
     }
 
+    /**
+     * This method is used to append the application side requested scopes after validating.
+     * The application can request the optional scopes either via adaptive scripts or via the authorize
+     * request query parameters.
+     * i.e. Adaptive Script example:
+     * This will ignore any other definition (common, local) of the authenticatorParams.
+     * var onLoginRequest = function(context) {
+     * executeStep(1, {
+     * authenticatorParams: {
+     * federated: {
+     * "Google Calender": {
+     * federated_token_scope: "https://googleapis.calander.readonly https://google.calander.list"
+     * }}}}, {});}
+     * i.e Authorize request query param example:
+     * /authorize?response_type=id_token&client_id={ClientId}&redirect_uri={https://app/callback}
+     * &scope=email profile openid
+     * &federated_token_scope=Google Calender;read write,Microsoft Authenticator;https://googleapis.calender
+     *
+     * @param context                 The authentication context.
+     * @param authenticatorProperties The authenticator properties.
+     * @param scopes                  The scopes defined in the authenticator properties.
+     * @return The IDP defined scope and the validated scopes requested by the application.
+     */
+    private String addValidScopesForFederatedTokenSharing(AuthenticationContext context,
+                                                          Map<String, String> authenticatorProperties, String scopes) {
+
+        // Get the application requested scopes for the federated tokens.
+        String requestedScopesForTokenSharing = getRequestedScopesForTokenSharing(context);
+
+        // Validating the application requested scopes by the authenticator allowed scopes for federated token sharing.
+        Set<String> validScopesForTokenSharing = validateScopeForTokenSharing(
+                authenticatorProperties.get(OIDCAuthenticatorConstants.FEDERATED_TOKEN_ALLOWED_SCOPE),
+                requestedScopesForTokenSharing);
+
+        if (CollectionUtils.isEmpty(validScopesForTokenSharing)) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("No matching scopes found for federated token sharing.");
+            }
+            return scopes;
+        }
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Valid scopes found for federated token sharing: " + validScopesForTokenSharing +
+                    ", IDP: " + getFederatedAuthenticatorName(context));
```

Comment:
shall we put the idp name 1st and then the scope list

Reply:
Fixed d2b40f9255b4a5ef3b8f9d44bd58ccdba9f50f70